module/purpose/Soccer/src/Soccer.cpp

```diff
@@ -269,22 +283,37 @@ namespace module::purpose {
         // A robot may emit leader messages mistakenly. Only listen to real leader.
         bool other_is_leader = active_robots.front().robot_id == incoming_robot_id;
         bool purposes_exist = !robocup.purposes.purposes.empty();
+        uint8_t striker_id = 0;
 
         if (other_is_leader && purposes_exist) {
+            // Find the striker by id
             for (const auto& purpose : robocup.purposes.purposes) {
-                if (purpose.player_id == PLAYER_ID ) {
-                    if (purpose.purpose == SoccerPosition::DEFENDER) {
-                        if (soccer_position != Position::DEFENDER) {
-                            emit<Task>(std::make_unique<Defender>(cfg.force_playing));
-                            log<NUClear::INFO>("Robot made defender");
-                        }
-                    }
-                    else if (purpose.purpose == SoccerPosition::STRIKER) {
-                        if (soccer_position != Position::STRIKER) {
-                            emit<Task>(std::make_unique<Striker>(cfg.force_playing));
-                            log<NUClear::INFO>("Robot made striker");
-                        }
-                    }
+                if (purpose.purpose == SoccerPosition::STRIKER){
+                    striker_id = purpose.player_id;
+                    break;
+                }
+            }
+
+            // Start our soccer position's task
+            if (striker_id == player_id) {
+                if (soccer_position != Position::STRIKER) {
+                    emit<Task>(std::make_unique<Striker>(cfg.force_playing));
+                    log<NUClear::INFO>("Robot made striker");
+                }
+            } else {
+                if (soccer_position != Position::DEFENDER) {
+                    emit<Task>(std::make_unique<Defender>(cfg.force_playing));
+                    log<NUClear::INFO>("Robot made defender");
+                }
+            }
+
+            // Store the new positions. Bit messy, but we should not reassign the striker if we become the leader later
+            // We cannot do this by index/same loop as our active_robots and our leader's active_robots may differ
+            for (int i = 0; i < int(active_robots.size()); ++i) {
```

Comment:
```suggestion
            for (size_t i = 0; i < active_robots.size(); ++i) {
```
something like this should work? or make `i` just whatever type `.size()` gives